pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java

```diff
@@ -265,7 +265,8 @@ public SendMessageInfo sendMessages(final List<Entry> entries, SendListener list
                 if (i == (entries.size() - 1)) {
                     promise = writePromise;
                 }
-                ctx.write(Commands.newMessage(consumerId, messageId, metadataAndPayload), promise);
+                int redeliveryCount = subscription.getDispatcher().getRedeliveryTracker().getRedeliveryCount(PositionImpl.get(messageId.getLedgerId(), messageId.getEntryId()));
```

Comment:
Instead of keeping a separate redelivery tracker, would it be possible to reuse the `pendingAcks` maps on the `Consumer` class?

`pendingAcks` is a map for which the key is a pair longs and the value is a pair of longs as well. 

`(ledgerId, entryId) --> (batchSize, none)`

As you can see, the second `long` field in the value is currently unused. We could keep the counter there.

Reply:
>My main concern with the new map is to have to potentially accumulate lots of items in the concurrent hash map, 

I'm understand your concern, i think the tracker map is limited by maxUnackMessages, only put element into tracker map when trigger by redeliverUnackedMessages().

>  irrespective of whether the DLQ is being used or not.

 I think we can optimize it next step, i think we can add a field in proto like `enableRedeliveryTracker`, broker just enable this mechanism when `enableRedeliveryTracker` is true.

> Especially, this might be mostly not the case in exclusive/failover subscriptions, where data is typically processed in order.

Yes, i'm agree, i think we can only enable the counter tracking only on Shared subscriptions, mean that just in PersistentDispatcherMultipleConsumers.